source/UIAHandler/remote.py

```diff
@@ -4,36 +4,150 @@
+	op = operation.Operation()
+
+	@op.buildFunction
+	def code(ra: remoteAPI.RemoteAPI):
+		remoteDocElement = ra.newElement(docElement)
+		remoteTextRange = ra.newTextRange(textRange)
+		remoteCustomAttribValue = ra.newVariant()
+		with ra.ifBlock(remoteDocElement.isExtensionSupported(guid_msWord_extendedTextRangePattern)):
+			ra.logRuntimeMessage("docElement supports extendedTextRangePattern")
+			remoteResult = ra.newVariant()
+			ra.logRuntimeMessage("doing callExtension for extendedTextRangePattern")
+			remoteDocElement.callExtension(
+				guid_msWord_extendedTextRangePattern,
+				remoteResult
+			)
+			with ra.ifBlock(remoteResult.isNull()):
+				ra.logRuntimeMessage("extendedTextRangePattern is null")
+				ra.halt()
+			with ra.elseBlock():
+				ra.logRuntimeMessage("got extendedTextRangePattern ")
+				remoteExtendedTextRangePattern = remoteResult.asType(RemoteExtensionTarget)
+				with ra.ifBlock(
+					remoteExtendedTextRangePattern.isExtensionSupported(guid_msWord_getCustomAttributeValue)
+				):
+					ra.logRuntimeMessage("extendedTextRangePattern supports getCustomAttributeValue")
+					ra.logRuntimeMessage("doing callExtension for getCustomAttributeValue")
+					remoteExtendedTextRangePattern.callExtension(
+						guid_msWord_getCustomAttributeValue,
+						remoteTextRange,
+						customAttribID,
+						remoteCustomAttribValue
+					)
+					ra.logRuntimeMessage("got customAttribValue of ", remoteCustomAttribValue)
+					ra.Return(remoteCustomAttribValue)
+				with ra.elseBlock():
+					ra.logRuntimeMessage("extendedTextRangePattern does not support getCustomAttributeValue")
+		with ra.elseBlock():
+			ra.logRuntimeMessage("docElement does not support extendedTextRangePattern")
+		ra.logRuntimeMessage("msWord_getCustomAttributeValue end")
+
+	customAttribValue = op.execute()
+	return customAttribValue
+
+
+def collectAllHeadingsInTextRange(
+	textRange: UIA.IUIAutomationTextRange
+) -> Generator[tuple[int, str, UIA.IUIAutomationElement], None, None]:
+	op = operation.Operation()
+
+	@op.buildIterableFunction
+	def code(ra: remoteAPI.RemoteAPI):
+		remoteTextRange = ra.newTextRange(textRange, static=True)
+		with remoteAlgorithms.remote_forEachUnitInTextRange(
+			ra, remoteTextRange, TextUnit.Paragraph
+		) as paragraphRange:
+			val = paragraphRange.getAttributeValue(AttributeId.StyleId)
+			with ra.ifBlock(val.isInt()):
+				intVal = val.asType(RemoteInt)
+				with ra.ifBlock((intVal >= StyleId.Heading1) & (intVal <= StyleId.Heading9)):
+					level = (intVal - StyleId.Heading1) + 1
+					label = paragraphRange.getText(-1)
+					ra.Yield(level, label, paragraphRange)
+
+	for level, label, paragraphRange in op.iterExecute(maxTries=20):
+		yield level, label, paragraphRange
+
+
+def findFirstHeadingInTextRange(
+	textRange: UIA.IUIAutomationTextRange,
+	wantedLevel: int | None = None,
+	reverse: bool = False
+) -> tuple[int, str, UIA.IUIAutomationElement] | None:
+	op = operation.Operation(enableCompiletimeLogging=True)

```

Comment:
Should compile-time logging be enabled for this once it's moved into a release build? Maybe this can be another logging category in the Advanced settings?